cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceService.java

```diff
@@ -122,15 +161,23 @@ public void setAuditPublisher(AuditPublisher auditPublisher) {
    * @throws NotFoundException if either the namespace or dataset instance is not found,
    * @throws IOException if there is a problem in making an HTTP request to check if the namespace exists.
    */
-  public DatasetMeta get(Id.DatasetInstance instance, List<? extends Id> owners) throws Exception {
+  public DatasetMeta get(final Id.DatasetInstance instance, List<? extends Id> owners) throws Exception {
+    // Application Deployment first makes a call to the dataset service to check if the instance already exists with
+    // a different type. To make sure that that call responds with the right exceptions if necessary, first fetch the
+    // meta from the cache and throw appropriate exceptions if necessary.
+    DatasetMeta datasetMeta;
     try {
-      return metaCache.get(instance);
+      datasetMeta = metaCache.get(instance);
     } catch (ExecutionException e) {
-      if ((e.getCause() instanceof Exception) && (e.getCause() instanceof HttpErrorStatusProvider)) {
-         throw (Exception) e.getCause();
+      Throwable cause = e.getCause();
+      if ((cause instanceof Exception) && (cause instanceof HttpErrorStatusProvider)) {
+         throw (Exception) cause;
       }
       throw e;
     }
+    // Only return the above datasetMeta if authorization succeeds
+    ensureAccess(instance.toEntityId());
```

Comment:
Added a comment at the beginning. The application deployment pipeline makes a getDataset call to figure out if it exists. Did it this way so that call can respond with the right error message instead of an authorization error.